### ISSUE TITLE
Block checkpointing service.

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#429]
Create a checkpointing block validation service, which validates blocks by checking their block header hashes against a list of known checkpoints. This service should not depend on chain state.

As an optimisation, the known checkpoint list should contain intermediate checkpoints of every Nth hash (N ~= 500). To accommodate chain reorgs, the final hash should be at least 100 hashes behind the tip (see #452).

Service Interface and Implementation
- [x] Create a checkpointing block validation service, which takes a block and a Network enum
  - [x] Work out how to turn a Network enum into a list of checkpoint hashes
    - [x] The initial stub can just use include_str!()
    - [ ] ~Use the config and a flat file?~

List Generation and Testing
- [x] Populate the checkpoint hash list using zcashd
- [x] Test the checkpoint hash list
  - [x] Validate the chain and warn on unused hashes
  - [x] Generate a fresh list using zcashd, and make sure the distributed list is a prefix of the fresh list

Further questions:
- [x] Do we need to add a "minimum chain work" fake chain check, like zcashd?
  - https://electriccoin.co/blog/new-release-2-1-0/
  - No, because we're already checkpointing to sapling.

**Comment by hdevalence** [comment]
To populate the checkpoint hash list, we can use a local copy of `zcashd`; the `zcash-cli` tool can be used to get block hashes like so:
```
./src/zcash-cli getblock 10 | jq '.hash' | tr -d '"'
```
The output needs to be passed through `zebrad revhex` since it's given in reversed order to what the Zebra internals expect.

I'm not exactly sure how things should be structured internally but I think it makes sense to separate the list of hashes into two lists, one a list of pre-Sapling hashes (whose last hash is the block at which Sapling activates) and the other a list of post-Sapling hashes.  The reason is that we're not planning to implement Sprout verification, so checkpointing on Sapling activation will be mandatory, while checkpoints post-Sapling should be optional.

**Comment by teor2345** [comment]
Added extra tests and questions based on zcashd's fake chain detection in https://electriccoin.co/blog/new-release-2-1-0/

**Comment by hdevalence** [comment]
Regarding the further questions: I don't think that we need to add a minimum chain work check, since we're already checkpointing all the way up to Sapling activation and that's an enormous amount of work to fake.

**Comment by hdevalence** [comment]
> Service Interface and Implementation
> 
> * [ ]  Create a checkpointing block validation service, which takes a block and a Network enum

Probably the `Network` enum can be passed to a constructor rather than in the `Request`, right?

>   * [ ]  Create a checkpoint list parsing service, which takes a Future of raw checkpoint list data     
>     * [ ]  Work out if we need to persistently store the parsed list in memory
>     * [ ]  Work out when we can drop the parsed list

I think it might work to store the checkpoint list in memory for the entire lifetime of the checkpointer, just dropping the entire checkpointer when it's no longer needed.  If it would work, it seems simpler.

> List Generation and Testing
>
> * [ ]  Make sure the node shuts down if it detects a fake chain: https://electriccoin.co/blog/new-release-2-1-0/

This might be possible to get as a side effect of a hybrid verifier design (more below).
 
> * [ ]  Do we want to distinguish between "skip validation" checkpoints and "validate, but require this block header hash" checkpoints? Or allow the checkpoint validator to operate in both modes?

One potential solution to this would be to create a (strawman)
```rust
pub struct HybridVerifier {
    checkpoint_verifier: CheckpointVerifier,
    block_verifier: BlockVerifier,
    ....
}
```
Then `HybridVerifier::call` could look at the block height and determine whether to verify with `checkpoint_verifier` or `block_verifier` or both, depending on the block height or some config settings.  This way all of the policy would be isolated from each verification method. At some later point we could change to an `Option<CheckpointVerifier>` to save on memory, but I'm not sure it's a big deal: at 32 bytes every 1000 blocks, keeping checkpoints up to 800,000 only costs 0.03MB.

This also means that we could factor out the "commit to the state" step from the individual verifiers.

> Here's how the "skip validation mode" could work:
> 
> ```
> max_checkpoint_height = max(checkpoint_list.heights())
> checkpoint_ancestor_list = checkpoint_list.copy()
> for every block where block.height() <= max_checkpoint_height:
>   hash = block.header_hash()
>   if hash in checkpoint_ancestor_list.hashes():
>     state.AddBlock(block)
>     checkpoint_ancestor_list.append(block.height() - 1, block.parent_header_hash)
>   else if hash != checkpoint_ancestor_list.get_hash_at_height(block.height()):
>     reject(block)
> ```
> 
> We can make this mode efficient by waiting on a hash future for each block height. That future should resolve when the child block checkpoint inserts the height and hash of its parent block. (This is the reverse of normal block validation, where blocks wait on the future for their entire parent block.)

Notes on some discussion with @teor2345 on Discord on how to do this concretely:

The `Ed25519Verifier` in `tower-batch/tests` is an example of a kind of similar problem, in that it collects signature items and then later uses a `broadcast` channel to send the verification result to all callers.  However, because we want to have the verification future resolve to the hash of the verified block, we probably need to use a oneshot, like (strawman):
```rust
// on creation, pass the rx end of the oneshot to the caller as the future
struct QueuedBlock {
    block: Arc<Block>,
    hash: BlockHeaderHash, // cached block.into()
    tx: tokio::sync::oneshot::Sender<Result<BlockHeaderHash, Error>>,
}
```

Thinking about how blocks flow into the checkpointer:

* we want to be able to stream blocks into the verifier and get results without having to worry about checkpoint intervals;
* we don't ever want to have gaps in the block chain, so even though we could verify the last block in a checkpoint interval immediately, we probably want to wait until we have a complete chain back to the previous checkpoint.
* we should try to handle the case where the caller attempts to verify multiple blocks at the same height.

To do this transparently, we could do something like (strawman):
```
// Should start with checkpoint_frequency empty vecs;
// vec at position i is the list of candidate blocks at height i within the interval
type Interval = Vec<Vec<QueuedBlock>>

struct CheckpointVerifier {
    checkpoint_frequency: usize,
    // key is block height / checkpoint_frequency
    queued: HashMap<usize, Vec<Vec<QueuedBlock>>>,
}
```
This would have a method to queue blocks for verification by putting them in the right slot in the hashmap and returning the rx end of the of the oneshot, and a method to check whether a particular hashmap entry is ready for verification. In order to not have gaps in the block chain, we'd need to first verify that there exists a path from the checkpoint to the beginning of the interval, before actually sending anything through any of the oneshots, so that we know we're accepting an entire interval at once.

If the `QueuedBlock` caches the hash of the current block as above, this check would be very cheap: start at the last entry in the interval and attempt to walk backwards until we hit an empty slot.  If we get all the way to the beginning, we can verify.  To handle forks (where `queued[h].len() > 1`), we could prefer to select the block referred to by the previous chain.

Alternatively, if we didn't care about supporting forks, we could use a `Vec<Option<QueuedBlock>>` and just immediately fail verification of whichever block came second, leaving the first one until we have a complete interval.  This would be simpler.  I'm not sure how to weigh the importance of handling forks, since the checkpointer is going to be used > 100 blocks behind the tip anyways.

**Comment by teor2345** [comment]
There is only ever one correct block at each checkpoint verifier height: the block whose descendant is the next checkpoint.

If all nodes are:
* synced near the latest tip,
* prune old side-chains, and
* only advertise the main chain,

we would never get any side-chains during checkpointing. But we can't guarantee any of these things, due to unreliable networks, node uptime, and implementation quirks.

Instead, the checkpoint verifier should keep a list of blocks at each height. When a particular height can chain forward to the next checkpoint, the verifier can discard all the side-chains.

**Comment by teor2345** [comment]
That said, the fanout design in #504 should allow us to recover from an ancient fork. If a checkpoint fails, we'll reload all its blocks, and almost always get the main chain the second time.

So I think I'll go with the simpler design, and add side-chain tests to make sure we can recover.

**Comment by teor2345** [comment]
This is effectively done, and we have tickets for the remaining work.
